s2spy/preprocess.py

```diff
@@ -0,0 +1,217 @@
+         - Calculate the rolling mean of the input data.
+         - Calculate and store the trend of the rolling mean.
+         - Calculate and store the climatology of the rolling mean.
+
+        When calling `.transform(data)`, the preprocessor will:
+         - Remove the (stored) trend from a copy of the data.
+         - Remove the climatology from this detrended data.
+         - Return the detrended and deseasonalized data.
+
+        Args:
+            rolling_window_size: The size of the rolling window that will be applied
+                before calculating the trend and climatology. Setting this to 1 will
+                effectively skip this step.
+            rolling_min_periods: The minimum number of periods within a rolling window.
+                If higher than 1 (the default), NaN values will be present at the start
+                and end of the preprocessed data.
+            detrend: Which method to use for detrending. Currently the only method
+                supported is "linear". If you want to skip detrending, set this to None.
+            remove_climatology (optional): If you want to calculate and remove the
+                climatology of the data. Defaults to True.
+        """
+        self._window_size = rolling_window_size
+        self._min_periods = rolling_min_periods
+        self._detrend = detrend
+        self._remove_climatology = remove_climatology
+
+        self._climatology: Union[xr.DataArray, xr.Dataset]
+        self._trend: dict
+        self._is_fit = False
+
+    def fit(self, data: Union[xr.DataArray, xr.Dataset]) -> None:
+        """Fit this Preprocessor to input data.
+
+        Args:
+            data
+        """
+        _check_input_data(data)
+        if self._window_size is not None:
+            data_rolling = data.rolling(
+                dim={"time": self._window_size}, min_periods=self._min_periods, center=True
+            ).mean()
+        # TODO: give option to be a gaussian-like window, instead of a block.
+        else:
+            data_rolling = data
+
+        if self._detrend is not None:
+            self._trend = _get_trend(data_rolling, self._detrend)
+
+        if self._remove_climatology:
+            self._climatology = _get_climatology(
+                (
+                    _apply_trend(data_rolling, self._detrend, self._trend)
+                    if self._detrend is not None
+                    else data_rolling
+                )
+            )
+        self._is_fit = True
+
+    def transform(
+        self, data: Union[xr.DataArray, xr.Dataset]
+    ) -> Union[xr.DataArray, xr.Dataset]:
+        """Apply the preprocessing steps to the input data.
+
+        Args:
+            data
+
+        Returns:
+            Preprocessed data.
+        """
+        if not self._is_fit:
+            raise ValueError("The preprocessor has to be fit to data before a transform"
+                             " can be applied")
+
+        if self._detrend is not None:
```

Comment:
Really nice and concise! Although the class docstring clearly describes the steps and the order, I think it would be nice to give some more user feedback on which steps were taken in what order. Now you just specify if you want to do rolling mean or detrend etc. but if I call transform it is not clear to me in what order the steps were taken. In our company code, we provide a plotting option for every step to show what happens during the transform like so: 
![image1](https://user-images.githubusercontent.com/82503135/219042651-ea17fb63-8dab-4761-8cfc-0dfe15cd7529.png)
It could also be some print statements like: detrending... or deseasonalizing.. just to know what is happening without looking at the code.

Reply:
I think that plotting modules do not fit well in this simple preprocessing module. The operations are so simple than it is quite trivial for a user to apply them separately and/or plot them before and after transformation, for example:
```py
data_array.plot()
data_array = preprocessor.transform(data_array)
data_array.plot()
```

As for print statements, they do not belong in (non-debugging) code. [You won't find a single one in sklearn, for example](https://github.com/scikit-learn/scikit-learn/search?q=print%28). If a user wants to know what is going on, they can look at:

- typing `help(function_name)`
- follow examples in the documentation
- look at the auto api reference
- (as a last resort) look at the code